src/server/tiered_storage.cc

```diff
@@ -212,11 +185,39 @@ bool TieredStorage::ShardOpManager::ReportFetched(EntryId id, string_view value,
   if (!modified && !cache_fetched_)
     return false;
 
+  // A workaround - to avoid polluting in-memory table by reads that go into a snapshot.
+  // It's not precise because we may handle reads coming from client requests.
+  // TODO: to revisit this when we rewrite it with more efficient snapshotting algorithm.
   if (SliceSnapshot::IsSnaphotInProgress())
     return false;
 
-  SetInMemory(get<OpManager::KeyRef>(id), value, segment);
-  return true;
+  auto key = get<OpManager::KeyRef>(id);
+  auto* pv = Find(key);
+  if (pv && pv->IsExternal() && segment == pv->GetExternalSlice()) {
+    Upload(key.first, value, segment.length, pv);
+    return true;
+  }
+
+  LOG(DFATAL) << "Internal error, should not reach this";
+  return false;
```

Comment:
Actually, I just thought - what if it runs during flushall?

Reply:
Chef Vlad cooking even when he is on holiday :nerd_face: :man_cook:

Reply:
Yeah Roman tags me for reviews 😭 (joking, I can handle them 😎 )

Reply:
I do not know. Let's put FATAL and find out